File: src/test/scala/tree/TreeSpec.scala
package tree

import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers
import org.scalatest.EitherValues

class TreeSpec extends AnyFlatSpec with Matchers with EitherValues:

  "Tree.get" should "fail because node is None" in {
    val result = Tree.get("somepath", None)
    result.isLeft shouldBe true
    result.left.value.message should include("node")
  }

  it should "fail because path is empty" in {
    val result = Tree.get("", Some(Node()))
    result.isLeft shouldBe true
    result.left.value.message should include("path")
  }

  it should "fail because path is not found" in {
    val result = Tree.get("nonexistent", Some(Node()))
    result.isLeft shouldBe true
    result.left.value.message should include("path not found")
  }

  it should "return expected result for testdata" in {
    val result = Tree.get("src/test/resources/testdata", Some(Node()))
    result.isRight shouldBe true

    val node = result.value
    node.name shouldBe "test/resources/testdata"

    // Should have one leaf (somefile.scala) at root level
    node.leafs.map(_.name) should contain("somefile.scala")

    // Should have one subdirectory node
    node.nodes should have size 1
    // currentPackage strips the first path component
    node.nodes.head.name shouldBe "test/resources/testdata/somedir"

    // The subdirectory should contain two scala files
    val somedir = node.nodes.head
    somedir.leafs.map(_.name) should contain allOf ("somefile.scala", "someotherfile.scala")

    // Verify paths are set correctly
    val rootLeaf = node.leafs.find(_.name == "somefile.scala").get
    rootLeaf.path shouldBe "src/test/resources/testdata/somefile.scala"

    val subdirLeafs = somedir.leafs
    subdirLeafs.foreach { l =>
      l.path should startWith("src/test/resources/testdata/somedir/")
    }
  }

  "Leaf.ast" should "fail because file does not exist" in {
    val leaf = Leaf(name = "missing.scala", path = "nonexistent.scala")
    val result = leaf.ast()
    result.isLeft shouldBe true
    result.left.value.message should include("file not found")
  }

  it should "return early because file is empty" in {
    val leaf = Leaf(
      name = "somefile.scala",
      path = "src/test/resources/testdata/somedir/somefile.scala"
    )
    val result = leaf.ast()
    result.isRight shouldBe true
    result.value.syntaxTree shouldBe None
  }

  it should "return a valid syntax tree" in {
    val leaf = Leaf(
      name = "somefile.scala",
      path = "src/test/resources/testdata/somefile.scala"
    )
    val result = leaf.ast()
    result.isRight shouldBe true
    result.value.syntaxTree shouldBe defined
  }

File: spec/tree_spec.rb
# frozen_string_literal: true

require_relative "../lib/tree"

RSpec.describe Tree do
  describe ".get" do
    it "raises NodeError when node is nil" do
      expect { Tree.get("somepath", nil) }.to raise_error(Tree::NodeError)
    end

    it "raises PathError when path is empty" do
      node = Tree::Node.new
      expect { Tree.get("", node) }.to raise_error(Tree::PathError)
    end

    it "raises Errno::ENOENT when path does not exist" do
      node = Tree::Node.new
      expect { Tree.get("nonexistent_path", node) }.to raise_error(Errno::ENOENT)
    end

    it "returns expected tree structure" do
      node = Tree::Node.new
      result = Tree.get("testdata", node)

      expect(result.name).to eq("testdata")

      expect(result.leafs.length).to eq(1)
      expect(result.leafs[0].name).to eq("somefile.rb")
      expect(result.leafs[0].path).to eq("testdata/somefile.rb")

      expect(result.nodes.length).to eq(1)
      expect(result.nodes[0].name).to eq("somedir")
      expect(result.nodes[0].leafs.length).to eq(2)
      expect(result.nodes[0].leafs.map(&:name)).to contain_exactly("somefile.rb", "someotherfile.rb")
      expect(result.nodes[0].leafs.map(&:path)).to contain_exactly(
        "testdata/somedir/somefile.rb",
        "testdata/somedir/someotherfile.rb"
      )
    end
  end

  describe Tree::Leaf do
    describe "#ast" do
      it "raises Errno::ENOENT when file does not exist" do
        leaf = Tree::Leaf.new(name: "missing.rb", path: "missing.rb")
        expect { leaf.ast }.to raise_error(Errno::ENOENT)
      end

      it "returns nil for an empty file" do
        leaf = Tree::Leaf.new(name: "somefile.rb", path: "testdata/somedir/somefile.rb")
        expect(leaf.ast).to be_nil
        expect(leaf.syntax_tree).to be_nil
      end

      it "parses a valid Ruby file into an AST" do
        leaf = Tree::Leaf.new(name: "somefile.rb", path: "testdata/somefile.rb")
        leaf.ast

        expect(leaf.syntax_tree).not_to be_nil
        expect(leaf.syntax_tree).to be_a(Parser::AST::Node)
      end
    end
  end

  describe Tree::Node do
    describe "#print" do
      it "outputs the tree structure" do
        node = Tree::Node.new
        Tree.get("testdata", node)

        output = StringIO.new
        node.print(io: output)

        text = output.string
        expect(text).to include("testdata")
        expect(text).to include("somefile.rb")
        expect(text).to include("somedir")
      end
    end
  end
end
